crates/typst-kit/src/package.rs

```diff
@@ -186,3 +203,80 @@ impl PackageStorage {
         })
     }
 }
+
+/// An iterator that deserializes its items lazily.
+struct LazyDeser<T, I> {
+    inner: I,
+    _phantom: PhantomData<T>,
+}
+
+trait LazyDeserExt: Sized {
+    /// Creates an iterator that returns deserialized items from the current
+    /// iterator.
+    fn lazy_deser<T>(self) -> LazyDeser<T, Self>;
+}
+
+impl<'a, I> LazyDeserExt for I
+where
+    I: Iterator<Item = &'a serde_json::Value>,
+{
+    fn lazy_deser<T>(self) -> LazyDeser<T, Self> {
+        LazyDeser { inner: self, _phantom: PhantomData }
+    }
+}
+
+impl<'a, T, I> Iterator for LazyDeser<T, I>
+where
+    T: DeserializeOwned,
+    I: Iterator<Item = &'a serde_json::Value>,
+{
+    type Item = Result<T, serde_json::Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let inner = self.inner.next()?;
+        Some(serde_json::from_value(inner.clone()))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn lazy_deser_index() {
+        let storage = PackageStorage::with_index(
+            None,
+            None,
+            Downloader::new("typst/test"),
```

Comment:
This test should normally never download anything, but maybe we want to have a stronger guarantee that it can't (unless it's fine to make network calls in tests) by introducing an abstraction over `Downloader`?

Reply:
network calls shouldn't occur, but I think it's fine since there aren't any. adding an abstraction just to have this as a static guarantee feels a little overkill.